subsys/bluetooth/controller/ll_sw/openisa/hal/RV32M1/ecb.c

```diff
@@ -17,39 +17,194 @@
+
+	/* After encrypt/decrypt req is sent to CAU3, the Host CPU will
+	 * execute WFE() until CAU3 signals task done by setting the event.
+	 */
+	handle.taskDone = kCAU3_TaskDonePoll;
+
+	/* Loads the key into CAU3's DMEM and expands the AES key schedule */
+	status = CAU3_AES_SetKey(CAU3, &handle, keyAes, sizeof(keyAes));
+	if (status != kStatus_Success) {
+		LOG_ERR("CAUv3 AES key set failed %d", status);
+		return;
+	}
+
+	status = CAU3_AES_Encrypt(CAU3, &handle, clear, cipher);
+	if (status != kStatus_Success) {
+		LOG_ERR("CAUv3 AES encrypt failed %d", status);
+		return;
+	}
+
+	if (cipher_text_le) {
+		/* STD e function outputs in MSB thus reverse for the (le) */
+		mem_rcopy(cipher_text_le, &cipher[0], sizeof(cipher));
+	}
+
+	if (cipher_text_be) {
+		memcpy(cipher_text_be, &cipher, sizeof(cipher));
+	}
 }
 
 u32_t ecb_encrypt_nonblocking(struct ecb *ecb)
 {
 	return 0;
 }
 
-#ifdef RV32M1_ECB_IMPL
-static void ecb_cleanup(void)
+void isr_ecb(void *param)
 {
 }
 
-static void ecb_cb(u32_t status, u8_t *cipher_be, void *context)
+/*
+ * Used by ULL as in this example:
+ * ltk[16] is copied from the HCI packet, preserving the LSO to MSO format
+ * skd[16] is copied from the PDU, preserving the LSO to MSO format
+ * calc the Session Key and retrieve the MSO to LSO format because that is how
+ * the PDU payload uses the encrypted data; MIC is also MSO to LSO.
+	ecb_encrypt(&conn->llcp_enc.ltk[0],
+				&conn->llcp.encryption.skd[0], NULL,
+				&lll->ccm_rx.key[0]);
+ */
+u32_t ecb_ut1(void)
 {
-}
+	/*
+	 * LTK = 0x4C68384139F574D836BCF34E9DFB01BF (MSO to LSO)
+	 * SKD = SKDm || SKDs
+	 * SKD (LSO to MSO)
+	 * :0x13:0x02:0xF1:0xE0:0xDF:0xCE:0xBD:0xAC
+	 * :0x79:0x68:0x57:0x46:0x35:0x24:0x13:0x02
+	 * SK = Encrypt(LTK, SKD)
+	 * SK (LSO to MSO)
+	 * :0x66:0xC6:0xC2:0x27:0x8E:0x3B:0x8E:0x05
+	 * :0x3E:0x7E:0xA3:0x26:0x52:0x1B:0xAD:0x99
+	 */
+	static const u8_t ltk_le[16] = {
+		0xbf, 0x01, 0xfb, 0x9d, 0x4e, 0xf3, 0xbc, 0x36,
+		0xd8, 0x74, 0xf5, 0x39, 0x41, 0x38, 0x68, 0x4c
+	};
+	static const u8_t skd_le[16] = {
+		0x13, 0x02, 0xF1, 0xE0, 0xDF, 0xCE, 0xBD, 0xAC,
+		0x79, 0x68, 0x57, 0x46, 0x35, 0x24, 0x13, 0x02
+	};
+	u8_t key_le[16] = {};
+	u8_t key_ref_le[16] = {
+		0x66, 0xC6, 0xC2, 0x27, 0x8E, 0x3B, 0x8E, 0x05,
+		0x3E, 0x7E, 0xA3, 0x26, 0x52, 0x1B, 0xAD, 0x99
+	};
+	u32_t status = kStatus_Success;
+	u8_t *key;
 
-#endif /* RV32M1_ECB_IMPL */
+	/* calc the Session Key and compare vs the ref_le in LSO format */
+	ecb_encrypt(ltk_le, skd_le, key = key_le, NULL);
 
-void isr_ecb(void *param)
-{
-}
+	if (memcmp(key_ref_le, key, 16)) {
+		printk("Failed session key unit test\n");
+		status = kStatus_Fail;
+	}
 
+	printk("Session key: %02x %02x %02x %02x %02x %02x %02x %02x "
+			"%02x %02x %02x %02x %02x %02x %02x %02x\n",
+			key[0], key[1], key[2], key[3],
+			key[4], key[5], key[6], key[7],
+			key[8], key[9], key[10], key[11],
+			key[12], key[13], key[14], key[15]);
+
+	return status;
+}
 
 u32_t ecb_ut(void)
 {
-	return 0;
```

Comment:
This unit test appears to be testing the cau3 lib. Is this needed here? Shouldn't it be testing the ecb_encrypt functions instead, possibly by mocking the cau3?

Should unittests be moved under tests/bluetooth?
@cvinayak Same goes for the ecb_ut for nordic, should this unittest be here? Is it run in CI?

Reply:
Good point. I will delete _ut() as it is superseded by _ut1() … And I will rename ut1() to ut().
BTW, I will be interested in hearing how the community will decide on proper place for small unit tests.

Reply:
> BTW, I will be interested in hearing how the community will decide on proper place for small unit tests.

I'm guessing these are unittests that only developers run. If they are part of CI, they should be moved to the test folder in my opinion.

Reply:
The code and the previous use of _ut() from Nordic's ecb.c made me believe the unit tests in this file are manually called by the developers by using the actual software stack/objects/flows instead of mocking it/them. My understanding could be wrong, though. I made this unit tests (this as well as those from radio.c) with this idea in mind.... that is too complex to mock the whole LLL/data flows and rather define an unit tests with the convenient prototype such that it could replace the function under test and be used inside the actual software.